### PR TITLE
[SYCL] Remove <iostream> from header and source files

### PR DESCRIPTION
**Pull request by raaiq1**
-Removed \<iostream\> headers from header and source files to best conform to LLVM Coding Standards : https://llvm.org/docs/CodingStandards.html#include-iostream-is-forbidden . 

-Substituted the removed headers with a proxy header which uses external linkage to resolve symbols in \<iostream\>

### CODE REVIEW

**Comment by aelovikov-intel** [comment]
Can you please add the following tests:

1) Compile just `#include <sycl/sycl.hpp>`, ensure that global objects aren't created (i.e. resulting obj file is empty).
2)  The following compiles cleanly
```
#include <iostream>
#include <sycl/sycl.hpp>
void foo() { std::cout << 42; }
```
3) The following compiles cleanly
```
#include <sycl/sycl.hpp>
#include <iostream>
void foo() { std::cout << 42; }
```

**Comment by raaiq1** [comment]
> Can you please add the following tests:
> 
> 1. Compile just `#include <sycl/sycl.hpp>`, ensure that global objects aren't created (i.e. resulting obj file is empty).
> 2. The following compiles cleanly
> 
> ```
> #include <iostream>
> #include <sycl/sycl.hpp>
> void foo() { std::cout << 42; }
> ```
> 
> 3. The following compiles cleanly
> 
> ```
> #include <sycl/sycl.hpp>
> #include <iostream>
> void foo() { std::cout << 42; }
> ```

I'm sorry, I don't understand what you mean by 'compile cleanly'

**Comment by aelovikov-intel** [comment]
That `clang -fsycl -c foo.cpp` finishes without warnings/errors.

File: sycl/test/basic_tests/no_iostream.cpp [hunk 1]
```diff
@@ -0,0 +1,9 @@
+// RUN: %clangxx -fsycl %s -S -emit-llvm -o- | FileCheck %s
+// CHECK-NOT: @_ZStL8__ioinit = internal global %"class.std::ios_base::Init"
```
Comment (aelovikov-intel) [inline sycl/test/basic_tests/no_iostream.cpp:2]: Were there any difficulties to produce empty output module and verify that?

**Review by cperkinsintel — DISMISSED** [review]
LGTM!!

**Comment by cperkinsintel** [comment]
failing CUDA test is known to be flaky.  ( https://github.com/intel/llvm-test-suite/issues/1103 )

File: sycl/include/sycl/detail/iostream_proxy.hpp [hunk 2]
```diff
@@ -0,0 +1,12 @@
+#ifndef IOSTREAM_PROXY_HPP
+#define IOSTREAM_PROXY_HPP 1
```
Comment (aelovikov-intel) [inline sycl/include/sycl/detail/iostream_proxy.hpp:2]: I think the project relies on `#pragma once` without include guards.

File: sycl/include/sycl/backend_types.hpp [hunk 3]
```diff
@@ -11,9 +11,9 @@
 #include <sycl/detail/defines.hpp>
 
 #include <fstream>
-#include <iostream>
 #include <istream>
 #include <string>
+#include <sycl/detail/iostream_proxy.hpp>
```
Comment (aelovikov-intel) [inline sycl/include/sycl/backend_types.hpp:16]: Here and in other files: can we move this include to the `sycl/*` group. I.e., in this file move it between lines 11 and 12.

**Review by aelovikov-intel — APPROVED** [review]

**Comment by raaiq1** [comment]
@intel/llvm-gatekeepers slight ping to merge commit. Thanks

**Review by againull — APPROVED** [review]

**Review by v-klochkov — APPROVED** [review]

**Comment by pvchupin** [comment]
@raaiq1, please follow up on windows post commit issue: https://github.com/intel/llvm/runs/7587694915?check_suite_focus=true

**Comment by aelovikov-intel** [comment]
I'm working on the build failure as @raaiq1 is on holiday today.

**Pull request merged**